src/Validation/ValidationURLProvider.php

```diff
@@ -0,0 +1,307 @@
+
+		/*
+		 * If 'Your homepage displays' is set to 'Your latest posts', include the homepage.
+		 */
+		if ( 'posts' === get_option( 'show_on_front' ) && $this->is_template_supported( 'is_home' ) ) {
+			$urls[] = [
+				'url'  => home_url( '/' ),
+				'type' => 'home',
+			];
+		}
+
+		$amp_enabled_taxonomies = array_filter(
+			get_taxonomies( [ 'public' => true ] ),
+			[ $this, 'does_taxonomy_support_amp' ]
+		);
+		$public_post_types      = get_post_types( [ 'public' => true ], 'names' );
+
+		// Include one URL of each template/content type, then another URL of each type on the next iteration.
+		for ( $i = $offset; $i < $this->limit_per_type + $offset; $i++ ) {
+			// Include all public, published posts.
+			foreach ( $public_post_types as $post_type ) {
+				$post_ids = $this->get_posts_that_support_amp( $this->get_posts_by_type( $post_type, $i, 1 ) );
+				if ( ! empty( $post_ids[0] ) ) {
+					$urls[] = [
+						'url'  => get_permalink( $post_ids[0] ),
+						'type' => $post_type,
+					];
+				}
+			}
+
+			foreach ( $amp_enabled_taxonomies as $taxonomy ) {
+				$taxonomy_links = $this->get_taxonomy_links( $taxonomy, $i, 1 );
+				$link           = reset( $taxonomy_links );
+				if ( ! empty( $link ) ) {
+					$urls[] = [
+						'url'  => $link,
+						'type' => $taxonomy,
+					];
+				}
+			}
+
+			$author_page_urls = $this->get_author_page_urls( $i, 1 );
+			if ( ! empty( $author_page_urls[0] ) ) {
+				$urls[] = [
+					'url'  => $author_page_urls[0],
+					'type' => 'author',
+				];
+			}
+		}
+
+		// Only validate 1 date and 1 search page.
+		$url = $this->get_date_page();
+		if ( $url ) {
+			$urls[] = [
+				'url'  => $url,
+				'type' => 'date',
+			];
+		}
+		$url = $this->get_search_page();
+		if ( $url ) {
+			$urls[] = [
+				'url'  => $url,
+				'type' => 'search',
+			];
+		}
+
+		return $urls;
+	}
+
+	/**
+	 * Gets whether the template is supported.
+	 *
+	 * @param string $template The template to check.
+	 * @return bool Whether the template is supported.
+	 */
+	public function is_template_supported( $template ) {
+		// If the --include argument is present in the WP-CLI command, this template conditional must be present in it.
+		if ( ! empty( $this->include_conditionals ) ) {
+			return in_array( $template, $this->include_conditionals, true );
+		}
+		if ( $this->include_unsupported ) {
+			return true;
+		}
+
+		$supportable_templates = AMP_Theme_Support::get_supportable_templates();
+
+		// Check whether this taxonomy's template is supported, including in the 'AMP Settings' > 'Supported Templates' UI.
+		return ! empty( $supportable_templates[ $template ]['supported'] );
+	}
+
+	/**
+	 * Gets the posts IDs that support AMP.
+	 *
+	 * By default, this only gets the post IDs if they support AMP.
+	 * This means that 'Posts' isn't deselected in 'AMP Settings' > 'Supported Templates'.
+	 * And 'Enable AMP' isn't unchecked in the post's editor.
+	 *
+	 * @param array $ids The post IDs to check for AMP support.
```

Comment:
```suggestion
	 * @param int[] $ids The post IDs to check for AMP support.
```

I just noticed that the typing information for `post_supports_amp()` and its deprecated cousin is off, as they accept `WP_Post[]|int[]`.